library/init/meta/converter/interactive.lean

```diff
@@ -94,20 +95,37 @@ meta def for (p : parse parser.pexpr) (occs : parse (list_of small_nat)) (c : it
 do (r, lhs, _) ← tactic.target_lhs_rhs,
    pat ← tactic.pexpr_to_pattern p,
    s   ← simp_lemmas.mk_default, -- to be able to use congruence lemmas @[congr]
-   (found, new_lhs, pr) ←
-     tactic.ext_simplify_core 1 {zeta := ff, beta := ff, single_pass := tt, eta := ff,
-                                 proj := ff, fail_if_unchanged := ff, memoize := ff} s
-       (λ u, return u)
-       (λ i s r p e, do
-         matched ← (tactic.match_pattern pat e >> return tt) <|> return ff,
-         guard matched,
-         if i ∈ occs then do
-           ⟨new_e, pr⟩ ← c.convert e r,
-           return (i+1, new_e, some pr, tt)
-         else return (i+1, e, none, tt))
-       (λ a s r p e, tactic.failed)
-       r lhs,
-  update_lhs new_lhs pr
+   st ← tactic.read,
+   (found, new_lhs, pr) ← tactic.ext_simplify_core
+     (success 1 st)  -- loop counter, and whether the conversion tactic failed
+     {zeta := ff, beta := ff, single_pass := tt, eta := ff, proj := ff,
+      fail_if_unchanged := ff, memoize := ff}
+     s
+     (λ u, return u)
+     (λ i s r p e,
+       match i with
+       | (success n s') :=
+         do matched ← (tactic.match_pattern pat e >> return tt) <|> return ff,
+            guard matched,
+            if n ∈ occs then 
+              (λ s,
+                match (c.convert e r) s with
+                | (success r s')     := success (success (n+1) s', r.fst, some r.snd, tt) s'
+                | (exception f p s') := success (exception f p s', e,     none,       tt) s'
+                end
+              )
```

Comment:
I can't help feeling there's a better way to do this - the success path is trivial to write with a `return`, I just can't work out how to get the failure details using `<|>` or similar.